Getting a weak-vtables warning with clang on ubuntu 22.04

### Description

I noticed that when ever I try to compile with clang ( 11 - 15 as I have installed) on with my normal warning flags on linux then I always ends up with a warning from the exceptions header about weak-vtables, I do know that some would say I have alot of warning flags activated but that is because that is what I saw some time ago that fmtlib uses and chose to use it myself also to catch errors and posible bad practices from my side

I do use CPM-cmake to pull in libraries purely because that is able to store them in a local cache so I don't have to pull the libraries over and over for each project and it does use the github release

### Reproduction steps

Tested with clang 11, 12, 13, 14, 15 on ubuntu jammy (22.04)

The shell command is taken from cmake log
```shell
/usr/bin/clang++-15  -I/home/buildclient/.cache/CPM/nlohmann_json/b3708972f6694fe462e4112e47aa04f10d2390b4/nlohmann_json/include -g -Werror -Wall -Wextra -pedantic -Wconversion -Wundef -Wdeprecated -Wweak-vtables -Wshadow -Wno-gnu-zero-variadic-macro-arguments -Wzero-as-null-pointer-constant -std=gnu++17 -MD -MT CMakeFiles/tanja84dk_testproject.dir/src/main.cpp.o -MF CMakeFiles/tanja84dk_testproject.dir/src/main.cpp.o.d -o CMakeFiles/tanja84dk_testproject.dir/src/main.cpp.o -c /buildtest/Cpp-Tanja84dk-DockerClient/src/main.cpp
```

I'm using [CPM-cmake](https://github.com/cpm-cmake/CPM.cmake) to pull in libraries and in essence its just pulling the 3.11.2 release from the release tab by tag

```CMakeList
cmake_minimum_required(VERSION 3.16)
project(tanja84dk_testproject LANGUAGES CXX VERSION 0.1.0
)

list(APPEND CMAKE_MODULE_PATH "${CMAKE_CURRENT_SOURCE_DIR}/cmake")

include(cmake/CPM.cmake)

include(CheckCXXCompilerFlag)

if (CMAKE_CXX_COMPILER_ID MATCHES "GNU")
  set(PEDANTIC_COMPILE_FLAGS -pedantic-errors -Wall -Wextra -pedantic
      -Wold-style-cast -Wundef
      -Wredundant-decls -Wwrite-strings -Wpointer-arith
      -Wcast-qual -Wformat=2 -Wmissing-include-dirs
      -Wcast-align
      -Wctor-dtor-privacy -Wdisabled-optimization
      -Winvalid-pch -Woverloaded-virtual
      -Wconversion -Wundef
      -Wno-ctor-dtor-privacy -Wno-format-nonliteral)
  if (NOT CMAKE_CXX_COMPILER_VERSION VERSION_LESS 4.6)
      set(PEDANTIC_COMPILE_FLAGS ${PEDANTIC_COMPILE_FLAGS}
         -Wno-dangling-else -Wno-unused-local-typedefs)
  endif ()
  if (NOT CMAKE_CXX_COMPILER_VERSION VERSION_LESS 5.0)
      set(PEDANTIC_COMPILE_FLAGS ${PEDANTIC_COMPILE_FLAGS} -Wdouble-promotion
          -Wtrampolines -Wzero-as-null-pointer-constant -Wuseless-cast
          -Wvector-operation-performance -Wsized-deallocation -Wshadow)
  endif ()
  if (NOT CMAKE_CXX_COMPILER_VERSION VERSION_LESS 6.0)
      set(PEDANTIC_COMPILE_FLAGS ${PEDANTIC_COMPILE_FLAGS} -Wshift-overflow=2
          -Wnull-dereference -Wduplicated-cond)
  endif ()
  set(WERROR_FLAG -Werror)
endif ()

if (CMAKE_CXX_COMPILER_ID MATCHES "Clang")
  set(PEDANTIC_COMPILE_FLAGS -Wall -Wextra -pedantic -Wconversion -Wundef
    -Wdeprecated -Wweak-vtables -Wshadow
    -Wno-gnu-zero-variadic-macro-arguments)
  check_cxx_compiler_flag(-Wzero-as-null-pointer-constant HAS_NULLPTR_WARNING)
  if (HAS_NULLPTR_WARNING)
    set(PEDANTIC_COMPILE_FLAGS ${PEDANTIC_COMPILE_FLAGS} -Wzero-as-null-pointer-constant)
  endif ()
  set(WERROR_FLAG -Werror)
endif ()

if (MSVC)
  set(PEDANTIC_COMPILE_FLAGS /W4)
  set(WERROR_FLAG /WX)
endif ()

set(tanja84dk_testproject_SOURCES
    src/main.cpp
)

set(tanja84dk_testproject_LIBRARIES
    nlohmann_json
)

CPMAddPackage(
    NAME            nlohmann_json
    GIT_REPOSITORY  https://github.com/nlohmann/json
    VERSION         3.11.2
    GIT_TAG         v3.11.2
)

add_executable(tanja84dk_testproject ${tanja84dk_testproject_SOURCES})

set_target_properties(tanja84dk_testproject
    PROPERTIES
        ARCHIVE_OUTPUT_DIRECTORY "${CMAKE_BINARY_DIR}/lib"
        LIBRARY_OUTPUT_DIRECTORY "${CMAKE_BINARY_DIR}/lib"
        RUNTIME_OUTPUT_DIRECTORY "${CMAKE_BINARY_DIR}/bin"
        CXX_STANDARD 17
    FOLDER
        Application
)

target_link_libraries(tanja84dk_testproject PRIVATE ${tanja84dk_testproject_LIBRARIES})

target_compile_options(tanja84dk_testproject
    PRIVATE
        ${WERROR_FLAG}
        ${PEDANTIC_COMPILE_FLAGS})

target_compile_features(tanja84dk_testproject
    PRIVATE
        cxx_std_17
)

### Expected vs. actual results

Expected it to build successfully specially because I didn't do anything specially in the small test but it warned/errored with the weak-vtables in the include/nlohmann/detail/exceptions.hpp

### Minimal code example

```Shell
#include <iostream>
#include <nlohmann/json.hpp>

int main() { return 0; }
```


### Error messages

```Shell
[main] Building folder: Cpp-Tanja84dk-DockerClient 
[build] Starting build
[proc] Executing command: /usr/bin/cmake --build /buildtest/Cpp-Tanja84dk-DockerClient/build --config Debug --target all --
[build] [1/2  50% :: 2.195] Building CXX object CMakeFiles/tanja84dk_testproject.dir/src/main.cpp.o
[build] FAILED: CMakeFiles/tanja84dk_testproject.dir/src/main.cpp.o 
[build] /usr/bin/clang++-15  -I/home/buildclient/.cache/CPM/nlohmann_json/b3708972f6694fe462e4112e47aa04f10d2390b4/nlohmann_json/include -g -Werror -Wall -Wextra -pedantic -Wconversion -Wundef -Wdeprecated -Wweak-vtables -Wshadow -Wno-gnu-zero-variadic-macro-arguments -Wzero-as-null-pointer-constant -std=gnu++17 -MD -MT CMakeFiles/tanja84dk_testproject.dir/src/main.cpp.o -MF CMakeFiles/tanja84dk_testproject.dir/src/main.cpp.o.d -o CMakeFiles/tanja84dk_testproject.dir/src/main.cpp.o -c /buildtest/Cpp-Tanja84dk-DockerClient/src/main.cpp
[build] In file included from /buildtest/Cpp-Tanja84dk-DockerClient/src/main.cpp:2:
[build] In file included from /home/buildclient/.cache/CPM/nlohmann_json/b3708972f6694fe462e4112e47aa04f10d2390b4/nlohmann_json/include/nlohmann/json.hpp:35:
[build] In file included from /home/buildclient/.cache/CPM/nlohmann_json/b3708972f6694fe462e4112e47aa04f10d2390b4/nlohmann_json/include/nlohmann/adl_serializer.hpp:14:
[build] In file included from /home/buildclient/.cache/CPM/nlohmann_json/b3708972f6694fe462e4112e47aa04f10d2390b4/nlohmann_json/include/nlohmann/detail/conversions/from_json.hpp:23:
[build] /home/buildclient/.cache/CPM/nlohmann_json/b3708972f6694fe462e4112e47aa04f10d2390b4/nlohmann_json/include/nlohmann/detail/exceptions.hpp:36:7: error: 'exception' has no out-of-line virtual method definitions; its vtable will be emitted in every translation unit [-Werror,-Wweak-vtables]
[build] class exception : public std::exception
[build]       ^
[build] /home/buildclient/.cache/CPM/nlohmann_json/b3708972f6694fe462e4112e47aa04f10d2390b4/nlohmann_json/include/nlohmann/detail/exceptions.hpp:134:7: error: 'parse_error' has no out-of-line virtual method definitions; its vtable will be emitted in every translation unit [-Werror,-Wweak-vtables]
[build] class parse_error : public exception
[build]       ^
[build] /home/buildclient/.cache/CPM/nlohmann_json/b3708972f6694fe462e4112e47aa04f10d2390b4/nlohmann_json/include/nlohmann/detail/exceptions.hpp:187:7: error: 'invalid_iterator' has no out-of-line virtual method definitions; its vtable will be emitted in every translation unit [-Werror,-Wweak-vtables]
[build] class invalid_iterator : public exception
[build]       ^
[build] /home/buildclient/.cache/CPM/nlohmann_json/b3708972f6694fe462e4112e47aa04f10d2390b4/nlohmann_json/include/nlohmann/detail/exceptions.hpp:205:7: error: 'type_error' has no out-of-line virtual method definitions; its vtable will be emitted in every translation unit [-Werror,-Wweak-vtables]
[build] class type_error : public exception
[build]       ^
[build] /home/buildclient/.cache/CPM/nlohmann_json/b3708972f6694fe462e4112e47aa04f10d2390b4/nlohmann_json/include/nlohmann/detail/exceptions.hpp:222:7: error: 'out_of_range' has no out-of-line virtual method definitions; its vtable will be emitted in every translation unit [-Werror,-Wweak-vtables]
[build] class out_of_range : public exception
[build]       ^
[build] /home/buildclient/.cache/CPM/nlohmann_json/b3708972f6694fe462e4112e47aa04f10d2390b4/nlohmann_json/include/nlohmann/detail/exceptions.hpp:239:7: error: 'other_error' has no out-of-line virtual method definitions; its vtable will be emitted in every translation unit [-Werror,-Wweak-vtables]
[build] class other_error : public exception
[build]       ^
[build] 6 errors generated.
[build] ninja: build stopped: subcommand failed.
[proc] The command: /usr/bin/cmake --build /buildtest/Cpp-Tanja84dk-DockerClient/build --config Debug --target all -- exited with code: 1
[driver] Build completed: 00:00:02.244
[build] Build finished with exit code 1
```


### Compiler and operating system

Clang 11, 12, 13, 14, 15 on ubuntu 22.04

### Library version

3.11.2

### Validation

- [X] The bug also occurs if the latest version from the [`develop`](https://github.com/nlohmann/json/tree/develop) branch is used.
- [ ] I can successfully [compile and run the unit tests](https://github.com/nlohmann/json#execute-unit-tests).

--- a/cmake/ci.cmake
+++ b/cmake/ci.cmake
@@ -96,7 +96,6 @@ file(GLOB_RECURSE SRC_FILES ${PROJECT_SOURCE_DIR}/include/nlohmann/*.hpp)
 # -Wno-padded                     We do not care about padding warnings.
 # -Wno-covered-switch-default     All switches list all cases and a default case.
 # -Wno-unsafe-buffer-usage        Otherwise Doctest would not compile.
-# -Wno-weak-vtables               The library is header-only.
 # -Wreserved-identifier           See https://github.com/onqtam/doctest/issues/536.
 
 set(CLANG_CXXFLAGS

--- a/cmake/ci.cmake
+++ b/cmake/ci.cmake
@@ -109,7 +109,6 @@ set(CLANG_CXXFLAGS
     -Wno-padded
     -Wno-covered-switch-default
     -Wno-unsafe-buffer-usage
-    -Wno-weak-vtables
     -Wno-reserved-identifier
 )
 

--- a/include/nlohmann/detail/exceptions.hpp
+++ b/include/nlohmann/detail/exceptions.hpp
@@ -47,6 +47,8 @@ class exception : public std::exception
     /// the id of the exception
     const int id; // NOLINT(cppcoreguidelines-non-private-member-variables-in-classes)
 
+    ~exception() override;
+
   protected:
     JSON_HEDLEY_NON_NULL(3)
     exception(int id_, const char* what_arg) : id(id_), m(what_arg) {} // NOLINT(bugprone-throw-keyword-missing)
@@ -131,6 +133,9 @@ class exception : public std::exception
     std::runtime_error m;
 };
 
+// out-of-line definition for exception to fix weak-vtables warning
+exception::~exception() = default;
+
 /// @brief exception indicating a parse error
 /// @sa https://json.nlohmann.me/api/basic_json/parse_error/
 class parse_error : public exception
@@ -173,6 +178,8 @@ class parse_error : public exception
     */
     const std::size_t byte;
 
+    ~parse_error() override;
+
   private:
     parse_error(int id_, std::size_t byte_, const char* what_arg)
         : exception(id_, what_arg), byte(byte_) {}
@@ -183,6 +190,9 @@ class parse_error : public exception
                       ", column ", std::to_string(pos.chars_read_current_line));
     }
 };
+
+// out-of-line definition for exception to fix weak-vtables warning
+parse_error::~parse_error() = default;
 
 /// @brief exception indicating errors with iterators
 /// @sa https://json.nlohmann.me/api/basic_json/invalid_iterator/
@@ -196,11 +206,16 @@ class invalid_iterator : public exception
         return {id_, w.c_str()};
     }
 
+    ~invalid_iterator() override;
+
   private:
     JSON_HEDLEY_NON_NULL(3)
     invalid_iterator(int id_, const char* what_arg)
         : exception(id_, what_arg) {}
 };
+
+// out-of-line definition for exception to fix weak-vtables warning
+invalid_iterator::~invalid_iterator() = default;
 
 /// @brief exception indicating executing a member function with a wrong type
 /// @sa https://json.nlohmann.me/api/basic_json/type_error/
@@ -214,10 +229,15 @@ class type_error : public exception
         return {id_, w.c_str()};
     }
 
+    ~type_error() override;
+
   private:
     JSON_HEDLEY_NON_NULL(3)
     type_error(int id_, const char* what_arg) : exception(id_, what_arg) {}
 };
+
+// out-of-line definition for exception to fix weak-vtables warning
+type_error::~type_error() = default;
 
 /// @brief exception indicating access out of the defined range
 /// @sa https://json.nlohmann.me/api/basic_json/out_of_range/
@@ -231,10 +251,15 @@ class out_of_range : public exception
         return {id_, w.c_str()};
     }
 
+    ~out_of_range() override;
+
   private:
     JSON_HEDLEY_NON_NULL(3)
     out_of_range(int id_, const char* what_arg) : exception(id_, what_arg) {}
 };
+
+// out-of-line definition for exception to fix weak-vtables warning
+out_of_range::~out_of_range() = default;
 
 /// @brief exception indicating other library errors
 /// @sa https://json.nlohmann.me/api/basic_json/other_error/
@@ -248,10 +273,15 @@ class other_error : public exception
         return {id_, w.c_str()};
     }
 
+    ~other_error() override;
+
   private:
     JSON_HEDLEY_NON_NULL(3)
     other_error(int id_, const char* what_arg) : exception(id_, what_arg) {}
 };
+
+// out-of-line definition for exception to fix weak-vtables warning
+other_error::~other_error() = default;
 
 }  // namespace detail
 NLOHMANN_JSON_NAMESPACE_END

--- a/include/nlohmann/detail/exceptions.hpp
+++ b/include/nlohmann/detail/exceptions.hpp
@@ -25,6 +25,18 @@
 #include <nlohmann/detail/meta/type_traits.hpp>
 #include <nlohmann/detail/string_concat.hpp>
 
+// With -Wweak-vtables, Clang will complain about the exception classes as they
+// have no out-of-line virtual method definitions and their vtable will be
+// emitted in every translation unit. This issue cannot be fixed with a
+// header-only library as there is no implementation file to move these
+// functions to. As a result, we suppress this warning here to avoid client
+// code to stumble over this. See https://github.com/nlohmann/json/issues/4087
+// for a discussion.
+#if defined(__clang__)
+    #pragma clang diagnostic push
+    #pragma clang diagnostic ignored "-Wweak-vtables"
+#endif
+
 NLOHMANN_JSON_NAMESPACE_BEGIN
 namespace detail
 {
@@ -178,8 +190,6 @@ class parse_error : public exception
     */
     const std::size_t byte;
 
-    ~parse_error() override;
-
   private:
     parse_error(int id_, std::size_t byte_, const char* what_arg)
         : exception(id_, what_arg), byte(byte_) {}
@@ -190,9 +200,6 @@ class parse_error : public exception
                       ", column ", std::to_string(pos.chars_read_current_line));
     }
 };
-
-// out-of-line definition for exception to fix weak-vtables warning
-parse_error::~parse_error() = default;
 
 /// @brief exception indicating errors with iterators
 /// @sa https://json.nlohmann.me/api/basic_json/invalid_iterator/
@@ -206,16 +213,11 @@ class invalid_iterator : public exception
         return {id_, w.c_str()};
     }
 
-    ~invalid_iterator() override;
-
   private:
     JSON_HEDLEY_NON_NULL(3)
     invalid_iterator(int id_, const char* what_arg)
         : exception(id_, what_arg) {}
 };
-
-// out-of-line definition for exception to fix weak-vtables warning
-invalid_iterator::~invalid_iterator() = default;
 
 /// @brief exception indicating executing a member function with a wrong type
 /// @sa https://json.nlohmann.me/api/basic_json/type_error/
@@ -229,15 +231,10 @@ class type_error : public exception
         return {id_, w.c_str()};
     }
 
-    ~type_error() override;
-
   private:
     JSON_HEDLEY_NON_NULL(3)
     type_error(int id_, const char* what_arg) : exception(id_, what_arg) {}
 };
-
-// out-of-line definition for exception to fix weak-vtables warning
-type_error::~type_error() = default;
 
 /// @brief exception indicating access out of the defined range
 /// @sa https://json.nlohmann.me/api/basic_json/out_of_range/
@@ -251,15 +248,10 @@ class out_of_range : public exception
         return {id_, w.c_str()};
     }
 
-    ~out_of_range() override;
-
   private:
     JSON_HEDLEY_NON_NULL(3)
     out_of_range(int id_, const char* what_arg) : exception(id_, what_arg) {}
 };
-
-// out-of-line definition for exception to fix weak-vtables warning
-out_of_range::~out_of_range() = default;
 
 /// @brief exception indicating other library errors
 /// @sa https://json.nlohmann.me/api/basic_json/other_error/
@@ -273,15 +265,14 @@ class other_error : public exception
         return {id_, w.c_str()};
     }
 
-    ~other_error() override;
-
   private:
     JSON_HEDLEY_NON_NULL(3)
     other_error(int id_, const char* what_arg) : exception(id_, what_arg) {}
 };
 
-// out-of-line definition for exception to fix weak-vtables warning
-other_error::~other_error() = default;
-
 }  // namespace detail
 NLOHMANN_JSON_NAMESPACE_END
+
+#if defined(__clang__)
+    #pragma clang diagnostic pop
+#endif

--- a/include/nlohmann/detail/exceptions.hpp
+++ b/include/nlohmann/detail/exceptions.hpp
@@ -59,8 +59,6 @@ class exception : public std::exception
     /// the id of the exception
     const int id; // NOLINT(cppcoreguidelines-non-private-member-variables-in-classes)
 
-    ~exception() override;
-
   protected:
     JSON_HEDLEY_NON_NULL(3)
     exception(int id_, const char* what_arg) : id(id_), m(what_arg) {} // NOLINT(bugprone-throw-keyword-missing)
@@ -144,9 +142,6 @@ class exception : public std::exception
     /// an exception object as storage for error messages
     std::runtime_error m;
 };
-
-// out-of-line definition for exception to fix weak-vtables warning
-exception::~exception() = default;
 
 /// @brief exception indicating a parse error
 /// @sa https://json.nlohmann.me/api/basic_json/parse_error/

--- a/single_include/nlohmann/json.hpp
+++ b/single_include/nlohmann/json.hpp
@@ -4426,8 +4426,6 @@ class exception : public std::exception
     /// the id of the exception
     const int id; // NOLINT(cppcoreguidelines-non-private-member-variables-in-classes)
 
-    ~exception() override;
-
   protected:
     JSON_HEDLEY_NON_NULL(3)
     exception(int id_, const char* what_arg) : id(id_), m(what_arg) {} // NOLINT(bugprone-throw-keyword-missing)
@@ -4511,9 +4509,6 @@ class exception : public std::exception
     /// an exception object as storage for error messages
     std::runtime_error m;
 };
-
-// out-of-line definition for exception to fix weak-vtables warning
-exception::~exception() = default;
 
 /// @brief exception indicating a parse error
 /// @sa https://json.nlohmann.me/api/basic_json/parse_error/

--- a/single_include/nlohmann/json.hpp
+++ b/single_include/nlohmann/json.hpp
@@ -4392,6 +4392,18 @@ inline OutStringType concat(Args && ... args)
 NLOHMANN_JSON_NAMESPACE_END
 
 
+// With -Wweak-vtables, Clang will complain about the exception classes as they
+// have no out-of-line virtual method definitions and their vtable will be
+// emitted in every translation unit. This issue cannot be fixed with a
+// header-only library as there is no implementation file to move these
+// functions to. As a result, we suppress this warning here to avoid client
+// code to stumble over this. See https://github.com/nlohmann/json/issues/4087
+// for a discussion.
+#if defined(__clang__)
+    #pragma clang diagnostic push
+    #pragma clang diagnostic ignored "-Wweak-vtables"
+#endif
+
 NLOHMANN_JSON_NAMESPACE_BEGIN
 namespace detail
 {
@@ -4545,8 +4557,6 @@ class parse_error : public exception
     */
     const std::size_t byte;
 
-    ~parse_error() override;
-
   private:
     parse_error(int id_, std::size_t byte_, const char* what_arg)
         : exception(id_, what_arg), byte(byte_) {}
@@ -4557,9 +4567,6 @@ class parse_error : public exception
                       ", column ", std::to_string(pos.chars_read_current_line));
     }
 };
-
-// out-of-line definition for exception to fix weak-vtables warning
-parse_error::~parse_error() = default;
 
 /// @brief exception indicating errors with iterators
 /// @sa https://json.nlohmann.me/api/basic_json/invalid_iterator/
@@ -4573,16 +4580,11 @@ class invalid_iterator : public exception
         return {id_, w.c_str()};
     }
 
-    ~invalid_iterator() override;
-
   private:
     JSON_HEDLEY_NON_NULL(3)
     invalid_iterator(int id_, const char* what_arg)
         : exception(id_, what_arg) {}
 };
-
-// out-of-line definition for exception to fix weak-vtables warning
-invalid_iterator::~invalid_iterator() = default;
 
 /// @brief exception indicating executing a member function with a wrong type
 /// @sa https://json.nlohmann.me/api/basic_json/type_error/
@@ -4596,15 +4598,10 @@ class type_error : public exception
         return {id_, w.c_str()};
     }
 
-    ~type_error() override;
-
   private:
     JSON_HEDLEY_NON_NULL(3)
     type_error(int id_, const char* what_arg) : exception(id_, what_arg) {}
 };
-
-// out-of-line definition for exception to fix weak-vtables warning
-type_error::~type_error() = default;
 
 /// @brief exception indicating access out of the defined range
 /// @sa https://json.nlohmann.me/api/basic_json/out_of_range/
@@ -4618,15 +4615,10 @@ class out_of_range : public exception
         return {id_, w.c_str()};
     }
 
-    ~out_of_range() override;
-
   private:
     JSON_HEDLEY_NON_NULL(3)
     out_of_range(int id_, const char* what_arg) : exception(id_, what_arg) {}
 };
-
-// out-of-line definition for exception to fix weak-vtables warning
-out_of_range::~out_of_range() = default;
 
 /// @brief exception indicating other library errors
 /// @sa https://json.nlohmann.me/api/basic_json/other_error/
@@ -4640,18 +4632,17 @@ class other_error : public exception
         return {id_, w.c_str()};
     }
 
-    ~other_error() override;
-
   private:
     JSON_HEDLEY_NON_NULL(3)
     other_error(int id_, const char* what_arg) : exception(id_, what_arg) {}
 };
 
-// out-of-line definition for exception to fix weak-vtables warning
-other_error::~other_error() = default;
-
 }  // namespace detail
 NLOHMANN_JSON_NAMESPACE_END
+
+#if defined(__clang__)
+    #pragma clang diagnostic pop
+#endif
 
 // #include <nlohmann/detail/macro_scope.hpp>
 

--- a/single_include/nlohmann/json.hpp
+++ b/single_include/nlohmann/json.hpp
@@ -4414,6 +4414,8 @@ class exception : public std::exception
     /// the id of the exception
     const int id; // NOLINT(cppcoreguidelines-non-private-member-variables-in-classes)
 
+    ~exception() override;
+
   protected:
     JSON_HEDLEY_NON_NULL(3)
     exception(int id_, const char* what_arg) : id(id_), m(what_arg) {} // NOLINT(bugprone-throw-keyword-missing)
@@ -4498,6 +4500,9 @@ class exception : public std::exception
     std::runtime_error m;
 };
 
+// out-of-line definition for exception to fix weak-vtables warning
+exception::~exception() = default;
+
 /// @brief exception indicating a parse error
 /// @sa https://json.nlohmann.me/api/basic_json/parse_error/
 class parse_error : public exception
@@ -4540,6 +4545,8 @@ class parse_error : public exception
     */
     const std::size_t byte;
 
+    ~parse_error() override;
+
   private:
     parse_error(int id_, std::size_t byte_, const char* what_arg)
         : exception(id_, what_arg), byte(byte_) {}
@@ -4550,6 +4557,9 @@ class parse_error : public exception
                       ", column ", std::to_string(pos.chars_read_current_line));
     }
 };
+
+// out-of-line definition for exception to fix weak-vtables warning
+parse_error::~parse_error() = default;
 
 /// @brief exception indicating errors with iterators
 /// @sa https://json.nlohmann.me/api/basic_json/invalid_iterator/
@@ -4563,11 +4573,16 @@ class invalid_iterator : public exception
         return {id_, w.c_str()};
     }
 
+    ~invalid_iterator() override;
+
   private:
     JSON_HEDLEY_NON_NULL(3)
     invalid_iterator(int id_, const char* what_arg)
         : exception(id_, what_arg) {}
 };
+
+// out-of-line definition for exception to fix weak-vtables warning
+invalid_iterator::~invalid_iterator() = default;
 
 /// @brief exception indicating executing a member function with a wrong type
 /// @sa https://json.nlohmann.me/api/basic_json/type_error/
@@ -4581,10 +4596,15 @@ class type_error : public exception
         return {id_, w.c_str()};
     }
 
+    ~type_error() override;
+
   private:
     JSON_HEDLEY_NON_NULL(3)
     type_error(int id_, const char* what_arg) : exception(id_, what_arg) {}
 };
+
+// out-of-line definition for exception to fix weak-vtables warning
+type_error::~type_error() = default;
 
 /// @brief exception indicating access out of the defined range
 /// @sa https://json.nlohmann.me/api/basic_json/out_of_range/
@@ -4598,10 +4618,15 @@ class out_of_range : public exception
         return {id_, w.c_str()};
     }
 
+    ~out_of_range() override;
+
   private:
     JSON_HEDLEY_NON_NULL(3)
     out_of_range(int id_, const char* what_arg) : exception(id_, what_arg) {}
 };
+
+// out-of-line definition for exception to fix weak-vtables warning
+out_of_range::~out_of_range() = default;
 
 /// @brief exception indicating other library errors
 /// @sa https://json.nlohmann.me/api/basic_json/other_error/
@@ -4615,10 +4640,15 @@ class other_error : public exception
         return {id_, w.c_str()};
     }
 
+    ~other_error() override;
+
   private:
     JSON_HEDLEY_NON_NULL(3)
     other_error(int id_, const char* what_arg) : exception(id_, what_arg) {}
 };
+
+// out-of-line definition for exception to fix weak-vtables warning
+other_error::~other_error() = default;
 
 }  // namespace detail
 NLOHMANN_JSON_NAMESPACE_END

--- a/tests/src/unit-regression2.cpp
+++ b/tests/src/unit-regression2.cpp
@@ -233,7 +233,7 @@ inline void from_json(const nlohmann::json& j, FooBar& fb) // NOLINT(misc-use-in
 struct for_3171_base // NOLINT(cppcoreguidelines-special-member-functions)
 {
     for_3171_base(const std::string& /*unused*/ = {}) {}
-    virtual ~for_3171_base() = default;
+    virtual ~for_3171_base();
 
     virtual void _from_json(const json& j)
     {
@@ -243,11 +243,16 @@ struct for_3171_base // NOLINT(cppcoreguidelines-special-member-functions)
     std::string str{}; // NOLINT(readability-redundant-member-init)
 };
 
+for_3171_base::~for_3171_base() = default;
+
 struct for_3171_derived : public for_3171_base
 {
     for_3171_derived() = default;
+    ~for_3171_derived() override;
     explicit for_3171_derived(const std::string& /*unused*/) { }
 };
+
+for_3171_derived::~for_3171_derived() = default;
 
 inline void from_json(const json& j, for_3171_base& tb) // NOLINT(misc-use-internal-linkage)
 {

--- a/tests/src/unit-regression2.cpp
+++ b/tests/src/unit-regression2.cpp
@@ -235,6 +235,32 @@ struct for_3171_base // NOLINT(cppcoreguidelines-special-member-functions)
     for_3171_base(const std::string& /*unused*/ = {}) {}
     virtual ~for_3171_base();
 
+    for_3171_base(const for_3171_base& other) // NOLINT(hicpp-use-equals-default,modernize-use-equals-default)
+        : str(other.str)
+    {}
+
+    for_3171_base& operator=(const for_3171_base& other)
+    {
+        if (this != &other)
+        {
+            str = other.str;
+        }
+        return *this;
+    }
+
+    for_3171_base(for_3171_base&& other) noexcept
+        : str(std::move(other.str))
+    {}
+
+    for_3171_base& operator=(for_3171_base&& other) noexcept
+    {
+        if (this != &other)
+        {
+            str = std::move(other.str);
+        }
+        return *this;
+    }
+
     virtual void _from_json(const json& j)
     {
         j.at("str").get_to(str);
@@ -251,7 +277,7 @@ struct for_3171_derived : public for_3171_base
     ~for_3171_derived() override;
     explicit for_3171_derived(const std::string& /*unused*/) { }
 
-    for_3171_derived(const for_3171_derived& other)
+    for_3171_derived(const for_3171_derived& other) // NOLINT(hicpp-use-equals-default,modernize-use-equals-default)
         : for_3171_base(other)
     {}
 

--- a/tests/src/unit-regression2.cpp
+++ b/tests/src/unit-regression2.cpp
@@ -249,6 +249,10 @@ struct for_3171_derived : public for_3171_base
 {
     for_3171_derived() = default;
     ~for_3171_derived() override;
+    for_3171_derived(for_3171_derived&&) noexcept = default;
+    for_3171_derived(const for_3171_derived&) = default;
+    for_3171_derived& operator=(const for_3171_derived&) = default;
+    for_3171_derived& operator=(for_3171_derived&&) = default;
     explicit for_3171_derived(const std::string& /*unused*/) { }
 };
 

--- a/tests/src/unit-regression2.cpp
+++ b/tests/src/unit-regression2.cpp
@@ -249,11 +249,33 @@ struct for_3171_derived : public for_3171_base
 {
     for_3171_derived() = default;
     ~for_3171_derived() override;
-    for_3171_derived(for_3171_derived&&) noexcept = default;
-    for_3171_derived(const for_3171_derived&) = default;
-    for_3171_derived& operator=(const for_3171_derived&) = default;
-    for_3171_derived& operator=(for_3171_derived&&) = default;
     explicit for_3171_derived(const std::string& /*unused*/) { }
+
+    for_3171_derived(const for_3171_derived& other)
+        : for_3171_base(other)
+    {}
+
+    for_3171_derived& operator=(const for_3171_derived& other)
+    {
+        if (this != &other)
+        {
+            for_3171_base::operator=(other); // Call base class assignment operator
+        }
+        return *this;
+    }
+
+    for_3171_derived(for_3171_derived&& other) noexcept
+        : for_3171_base(std::move(other))
+    {}
+
+    for_3171_derived& operator=(for_3171_derived&& other) noexcept
+    {
+        if (this != &other)
+        {
+            for_3171_base::operator=(std::move(other)); // Call base class move assignment operator
+        }
+        return *this;
+    }
 };
 
 for_3171_derived::~for_3171_derived() = default;

